src/StructureFactors/ElementContraction.jl

```diff
@@ -19,41 +21,60 @@ struct FullTensor{NCorr} <: Contraction{SVector{NCorr, ComplexF64}} end
 ################################################################################
 # Constructors
 ################################################################################
+Trace(swt::SpinWaveTheory) = Trace([1,5,9])
+
 function Trace(sf::StructureFactor{N}) where {N}
     # Collect all indices for matrix elements 𝒮^αβ where α=β
     indices = Int64[]
-    for (ci, idx) in sf.idxinfo
-        α, β = ci.I
-        if α == β
-            push!(indices, idx)
+    for (ki,i) = sf.observable_ixs
+        autocorrelation_index = CartesianIndex(i,i)
+        if haskey(sf.correlations,autocorrelation_index)
+            push!(indices,sf.correlations[autocorrelation_index])
+        else
+            problematic_correlation = ki
+            error("Can't calculate trace because auto-correlation of the $problematic_correlation observable was not computed.")
         end
     end
-    # Check that there are the correct number of such elements
-    if N == 0 || sf.dipole_corrs
-        if length(indices) != 3
-            error("Not all diagonal elements of the structure factor have been computed. Can't calculate trace.")
-        end
-    else
-        if length(indices) != N*N-1
-            error("Not all diagonal elements of the structure factor have been computed. Can't calculate trace.")
-        end
+
+    # SQ N.B.: This error doesn't make much sense, does it?
+    # So what if they used a different number from the default number of observables?
+    # Case in point: If you are doing dipole correlations in SU(N) mode, you're not taking
+    # the full trace, and this will error out.
```

Comment:
Maybe I'm not following. If you're doing dipole correlations in SU(N) mode, it will take the first branch (since sf.dipole_corrs will be true). (The checking for 3 is not necessary, however, when sf.dipole_corrs is true.)

I guess there is ambiguity here in terms of what is meant by "trace." Basically, "trace" means Sxx+Syy+Szz unless the user has specified custom observables/correlations, in which case it means \sum Taa where a runs from 1 to N^2-1. 

My feeling was that there should be some check here to make sure your not just summing some random collection of diagonal elements. But I guess we expect a user specifying arbitrary correlations to be sophisticated enough to know what they're doing.

Reply:
(Although I guess the docs need to build first.)

Reply:
If you're in SU(N) mode, you may still have a number different from N*N-1 number of correlations included in your trace, for example if you have M observables, you have M =/= N*N-1 terms.

The code now just directly checks that, for each observable in the structure factor, that observable's autocorrelation is included in the trace.

Reply:
Indeed, the point of the check was precisely to prevent the user from thinking they're getting a trace if M is not equal to the dimension of the Lie algebra su(N), namely N^2-1. This has physical significance. (The check is quick and dirty in the sense that it doesn't check is the N^2-1 operators indeed form an orthonormal basis for Hermitian matrices.)

I'm not sure this should be called `Trace` if it's just a sum or some available set of autocorrelations, and I do think there should be a trace option. Maybe `Trace` should just mean Sxx+Syy+Szz, which is what most people will expect. If someone wants to do a trace of a full set of SU(N) observables, they'll just have to write their own formula.

Reply:
> The code now just directly checks that, for each observable in the structure factor, that observable's autocorrelation is included in the trace.

I guess this is fine too, now that I read it a bit more carefully, since, in the default situation, the user will get Sxx+Syy+Szz. In any case, just explaining the reasoning behind the check.